File: packages/ios/fixtures/zig-slice/shim.m
// A stand-in for the host shim, so the hand-off path can be exercised.
//
// Written in Objective-C rather than Swift on purpose. The contract Zig
// depends on is a runtime one — `objc_getClass("CraftSwiftShim")` plus a
// message send — and a `@objc` Swift class presents exactly that same
// contract. Using ObjC here keeps the harness to a single clang invocation
// with no swiftc, while testing the identical mechanism.
//
// The shape being proven is the important part: the shim answers *what*, never
// *how*. It produces a payload and hands it back through
// `craft_ios_deliver_result`, so the wire format, the request id and the
// escaping all stay on the Zig side. Two components replying to one page by
// two different routes is how this codebase accumulated five envelopes.
#import <Foundation/Foundation.h>

extern void craft_ios_deliver_result(const char *action, unsigned long action_len,
                                     const char *json, unsigned long json_len,
                                     long long request_id);

@interface CraftSwiftShim : NSObject
@end

@implementation CraftSwiftShim

+ (BOOL)handleAction:(NSString *)action
             payload:(NSString *)payload
           requestId:(long long)requestId {
    // Exactly one action, to prove the seam. A real shim dispatches the
    // ~105 the Swift template still owns.
    if (![action isEqualToString:@"hostOnlyPing"]) {
        return NO;  // Not ours. Zig answers with UnknownAction.
    }

    NSString *json = [NSString stringWithFormat:
        @"{\"servedBy\":\"host-shim\",\"echo\":%@}",
        payload.length ? [NSString stringWithFormat:@"\"%lu\"", (unsigned long)payload.length] : @"null"];

    const char *actionBytes = action.UTF8String;
    const char *jsonBytes = json.UTF8String;
    craft_ios_deliver_result(actionBytes, strlen(actionBytes),
                             jsonBytes, strlen(jsonBytes),
                             requestId);
    return YES;
}

@end

File: packages/ios/fixtures/zig-slice/shim.swift
// The host shim, mirroring the design `CraftApp.swift`'s CraftSwiftShim uses —
// including the part most likely to fail silently in production: discovering
// Zig's delivery exports with `dlsym` rather than linking them.
//
// That discovery is the template's load-bearing assumption. A pure-Swift app
// has no Zig runtime, so the template cannot use `@_silgen_name` (a hard link
// dependency); it resolves the symbols at runtime and declines when absent.
// Whether `dlsym(dlopen(nil, RTLD_NOW), ...)` finds a symbol that was
// *statically linked* into the executable is exactly the kind of fact that
// must be proven on a simulator, because if a build setting strips it from the
// dynamic symbol table the seam degrades to "shim declines everything" with no
// error anywhere.
//
// Two actions, proving the two reply routes:
//   hostOnlyPing  -> craft_ios_deliver_result  (the page's promise resolves)
//   hostOnlyFail  -> craft_ios_deliver_error   (the page's promise rejects)
// The error route matters independently: delivering a rejection as a result
// would run the app's then-branch with an error-shaped object.
import Foundation

private typealias DeliverResultFn = @convention(c) (
    UnsafePointer<CChar>, UInt, UnsafePointer<CChar>, UInt, Int64
) -> Void
private typealias DeliverErrorFn = @convention(c) (
    UnsafePointer<CChar>, UInt, UnsafePointer<CChar>, UInt,
    UnsafePointer<CChar>, UInt, Int64
) -> Void

private let deliverResult: DeliverResultFn? = {
    guard let sym = dlsym(dlopen(nil, RTLD_NOW), "craft_ios_deliver_result") else { return nil }
    return unsafeBitCast(sym, to: DeliverResultFn.self)
}()

private let deliverError: DeliverErrorFn? = {
    guard let sym = dlsym(dlopen(nil, RTLD_NOW), "craft_ios_deliver_error") else { return nil }
    return unsafeBitCast(sym, to: DeliverErrorFn.self)
}()

@objc(CraftSwiftShim)
public class CraftSwiftShim: NSObject {
    @objc public static func handleAction(
        _ action: String,
        payload: String,
        requestId: Int64
    ) -> Bool {
        switch action {
        case "hostOnlyPing":
            guard let deliverResult else { return false }
            let json = "{\"servedBy\":\"host-shim\",\"language\":\"swift\",\"payloadBytes\":\(payload.utf8.count)}"
            action.withCString { a in
                json.withCString { j in
                    deliverResult(a, UInt(strlen(a)), j, UInt(strlen(j)), requestId)
                }
            }
            return true

        case "hostOnlyFail":
            guard let deliverError else { return false }
            // The message deliberately contains the characters Swift's old
            // hand-escaping mangled: a backslash, a quote, and a newline. If
            // they reach the page intact, escaping lives on the Zig side where
            // it belongs.
            let message = "declined \\ \"on purpose\"\nsecond line"
            action.withCString { a in
                message.withCString { m in
                    "HOST_DECLINED".withCString { c in
                        deliverError(a, UInt(strlen(a)), m, UInt(strlen(m)),
                                     c, UInt(strlen(c)), requestId)
                    }
                }
            }
            return true

        default:
            return false
        }
    }
}
